File: Dynamic Programming/1335.Minimum Difficulty of a Job Schedule/Solution.py
class Solution:
    def minDifficulty(self, jobDifficulty: List[int], d: int) -> int:
        n = len(jobDifficulty)
        f = [[inf] * (d + 1) for _ in range(n + 1)]
        f[0][0] = 0
        for i in range(1, n + 1):
            for j in range(1, d + 1):
                x = 0
                for k in range(i - 1, -1, -1):
                    x = max(x, jobDifficulty[k])
                    f[i][j] = min(f[i][j], f[k][j - 1] + x)
        return -1 if f[n][d] >= inf else f[n][d]

File: Dynamic Programming/1335.Minimum Difficulty of a Job Schedule/Solution.java
class Solution {
    public int minDifficulty(int[] jobDifficulty, int d) {
        int n = jobDifficulty.length;
        int[][] f = new int[n + 1][d + 1];
        int inf = 0x3f3f3f3f;
        for (int[] r : f) {
            Arrays.fill(r, inf);
        }
        f[0][0] = 0;
        for (int i = 1; i < n + 1; i++) {
            for (int j = 1; j < d + 1; j++) {
                int x = 0;
                for (int k = i - 1; k > -1; k--) {
                    x = Math.max(x, jobDifficulty[k]);
                    f[i][j] = Math.min(f[i][j], f[k][j - 1] + x);
                }
            }
        }
        return f[n][d] >= inf ? -1 : f[n][d];
    }
}
